Support TopologyManager in Large NUMA System (NVIDIA GB200)

fanzhangio opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
Bug Fix 
/kind bug

#### What this PR does / why we need it:
Kubelet’s TopologyManager doesn’t scale well on systems with many NUMA nodes. On NVIDIA GB200, for example, there are 34 NUMA nodes but only 2 of them have CPUs suitable for scheduling pods. When a device plugin advertises all 34 NUMA nodes, TopologyManager’s generateDeviceTopologyHints ends up iterating over all subsets of that 34-bit mask (O(2^n)). With n=34, this becomes effectively unbounded, so kubelet never reaches the Allocate call. As a result, the virt-launcher pod stays in Pending and the VM remains stuck in Starting indefinitely.

This PR constrains the search space in two places:
- **Device-manager hints:** 
   Hint enumeration is limited to NUMA nodes that actually host available/reusable devices, and an “any-NUMA” fallback hint is added. This only reduces complexity when devices are located on a small subset of NUMA nodes. If a device plugin still reports devices spread across all N NUMA nodes, the candidate set remains size N and IterateBitMasks is still O(2^N). However, on realistic multi-NUMA systems (including GB200), GPUs are typically present only on a few CPU-bearing nodes, so the candidate set remains small and the search is tractable.
- **Memory-manager hints:** 
   NUMA nodes without CPUs are skipped (unless they already hold assignments), so hint enumeration is effectively limited to the 2 CPU nodes on GB200 and does not explode combinatorially.

#### Which issue(s) this PR is related to:

Fixes #135541
<!--
Please link relevant issues to help with tracking.

To automatically close the linked issue(s) when this PR is merged,
add the word "Fixes" before the issue number or link.
Do not use "Fixes" if the PR is of kind `failing-test` or `flake`.

Reference KEPs when applicable in addition to specific issues.


Examples:
Fixes #<issue number>
<issue link> (issue in a different repository)


If there is no associated issue, then write "N/A".
-->

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
KEP: https://github.com/kubernetes/enhancements/issues/5726

Tested in GB200 (with KubeVirt)

1. Set topology manager policy with large NUMA nodes in kubelet config 
```yaml
topologyManagerPolicy: restricted
topologyManagerScope: pod
topologyManagerPolicyOptions:
  prefer-closest-numa-nodes: "true"
  max-allowable-numa-nodes: "34"
```

2. Deployed Device plugin, KubeVirt etc.

3. Create Pod (in this test example, it is KubeVirt VMI) with VFIO device passthrough

4. kubelet topology manager successfully passthrough admission and have the pod scheduled. Allocate was successfully.
```yaml
NAME                                      READY   STATUS    RESTARTS   AGE   IP           NODE           NOMINATED NODE   READINESS GATES
virt-launcher-vmi-gb200-4gpu-numa-xntbt   2/2     Running   0          68s   10.244.7.5   gb200-node-4   <none>           1/1

NAME                  AGE   PHASE     IP           NODENAME       READY   LIVE-MIGRATABLE   PAUSED
vmi-gb200-4gpu-numa   81s   Running   10.244.7.5   gb200-node-4   True    False
```
<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

PiotrProkop reviewed (commented):
Please split the change into 2 commits, one for memorymanager and second for devicemanager.

ffromani commented:
/triage accepted
/priority important-longterm

fanzhangio commented:
Thanks @PiotrProkop @ffromani for the comments. I updated the PR. Please let me know if it makes sense, and I will split it into 3 commits for cpu-manager, memory-manager and device-manager separately.

fanzhangio commented on pkg/kubelet/cm/memorymanager/policy_static.go:
```diff
@@ -575,6 +586,25 @@ func (p *staticPolicy) calculateHints(machineState state.NUMANodeMap, pod *v1.Po
+func (p *staticPolicy) selectableNUMANodes(machineState state.NUMANodeMap) []int {
+	numaNodes := make([]int, 0, len(machineState))
+	for nodeID, numaState := range machineState {
+		if !p.nodeHasCPUs(nodeID) && numaState.NumberOfAssignments == 0 {
```
Ok. Dropped the NumberOfAssignments guard so memory-manager hint generation now simply skips NUMA nodes without CPUs for all cases.

fanzhangio committed:
cpumanager: filter hints to NUMA nodes with memory

Record NUMANodesWithMemory from cAdvisor topology and limit CPU topology hints to nodes that have both CPUs and memory, falling back to prior behavior when memory info is unavailable. Adjust topology hint tests to cover asymmetric NUMA layouts and update cmp options for cpuset.

fanzhangio committed:
devicemanager: constrain topology hints to device NUMA nodes

Limit device-manager topology hints to NUMA nodes that host available or reusable devices while preserving the any-NUMA fallback hint, and add unit tests for asymmetric NUMA layouts.

fanzhangio committed:
memorymanager: skip CPU-less NUMA nodes in hints

Have the static policy ignore NUMA nodes without CPUs when enumerating topology hints unless they already host allocations, preventing combinatorial explosion on asymmetric NUMA topologies. Add unit coverage for large NUMA layouts.

PiotrProkop commented:
LGTM, @klueska can you review?

klueska commented on pkg/kubelet/cm/cpumanager/policy_static.go:
```diff
@@ -669,6 +669,21 @@ func (p *staticPolicy) GetPodTopologyHints(logger logr.Logger, s state.State, po
+	if numaNodesWithMemory.Size() == 0 {
+		return numaNodes
+	}
+	return numaNodesWithMemory
```
Is there ever a case where `numaNodesWithMemory != numaNodes`? If there is, then let's continue the discussion on how this commit could be simplified. If there's not, I'm not convinced we need any changes to the `cpumanager` at all. My understanding is that we would never have a NUMA node associated with a CPU if it didn't also have memory attached to it.

klueska commented on pkg/kubelet/cm/memorymanager/policy_static.go:
```diff
@@ -575,6 +586,25 @@ func (p *staticPolicy) calculateHints(machineState state.NUMANodeMap, pod *v1.Po
+func (p *staticPolicy) nodeHasCPUs(nodeID int) bool {
+	if len(p.numaNodesWithCPUs) == 0 {
+		return true
+	}
```
I don't think we should do this. If there is some weird esoteric hardware that would require this, then the `TopologyManager` should not be used. We can advise them to just disable it. We should add a follow.up to improve the docs around this.

klueska reviewed (commented):
@fanzhangio, thanks for being patient while I took the time to look through this PR in detail.

Unfortunately, I don't think we can make this change without a KEP. When first discussing this PR on slack, I thought this problem could be cast as a bug fix for systems that online a bunch of "unusable" NUMA nodes from the perspective of the `TopologyManager`, but after careful consideration (and review of this PR), I don't think that's true anymore

Instead, I think we need to add a mode of operation where all hint providers only consider NUMA nodes where CPUs are attached. For the CPU manager this is already true (by definition), and for the MemoryManager and DeviceManager this would need to be protected by some sort of flag, so as not to break any existing use-cases.

This should take the form of a `TopologyManager` option, that gets propagated to all hint providers so that they only consider NUMA nodes that match this criteria.

I am happy to review / champion this KEP for the 1.36 time frame. Please feel free to ping me once you have something ready.

klueska commented:
Actually, one thing occurred to me that could be actionable now (and backportable as a bug fix to older k8s versions).

The topologymanager and its hint providers should **all** be honoring the `--reserved-memory` flag the same way that the memory manager already does. This flag is intended to make regions of memory from specific NUMA nodes invisible to the kubelet. When the full memory region of a given NUMA node is reserved, that NUMA node should be taken out of consideration by **all** hint providers, not just the memory manager. Currently they are not, and this is a bug in my opinion.
